### ISSUE TITLE
Trailing zeros in bytes truncated in binary data using python http client

### ISSUE THREAD

**Issue by tmorgansl** [triton-inference-server/server#2545]
**Description**
We have a model which expects binary data as its input which we would like to pass in via the http API using the binary tensor data extension. The http python client you have for the API supports this, however at the moment it expects the binary to be passed in using a numpy array with the `dtype` of `np.bytes_`.

The problem with this `dtype` is that it essentially translates to a string, and therefore removes trailing zeros:-

```
>>> import numpy as np
>>> a = np.array([b'\x00\x00'])
>>> a[0]
b''
```

This is intended behaviour and not a bug in numpy (see https://github.com/numpy/numpy/issues/3878). The correct thing to use in this situation is np.void:-

```
>>> a = np.array([b'\x00\x00'], dtype=np.void)
>>> a[0]
void(b'\x00\x00')
```

Your client doesn't seem to support this dtype however, and blows up.

**Triton Information**
20.12

Are you using the Triton container or did you build it yourself? The container 

**To Reproduce**
1. Setup triton with a model that expects binary data
2. Build the python http client
3. Pass in an array of bytes with trailing zeros, e.g.

```
server_url = ...
model_name = ...
model_version = ...

def make_rest_call(req: List[bytes]) -> None:
    i = tritonclient.InferInput('examples', [1,1], 'BYTES')
    arr = np.array([req], dtype=np.void) # <- setting the dtype is key here
    i.set_data_from_numpy(arr, binary_data=True)
    with tritonclient.InferenceServerClient(server_url) as server:
        resp = server.infer(model_name, [i], model_version)
``` 

The request fails because the dtype is unexpected

**Expected behavior**
The request is successfully parsed and sent to the server

**Comment by Tabrizian** [comment]
@tmorgansl Thanks for creating this issue. We are currently aware of this issue and working on a fix.

I'm just curious how were you able to solve it? The problem with `np.void` is that it does not work for numpy arrays with variable length in different dimensions. For example, the string below will not work:

```
np.array([b'123', b'1234'], dtype=np.void)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Invalid type promotion with void datatypes of different lengths. Use the `np.bytes_` datatype instead to pad the shorter value with trailing zero bytes.

```

**Comment by tmorgansl** [comment]
@Tabrizian Ah I see, for my use case all of the binaries are of fixed length so I did not see this issue.

Strangely, with my version of numpy (1.18.5) I do not see that exception, it pads the bytes instead:-

```
>>> import numpy as np
>>> np.array([b'123', b'1234'], dtype=np.void)
array([b'\x31\x32\x33\x00', b'\x31\x32\x33\x34'], dtype='|V4')
```

but that would certainly break as well

**Comment by Tabrizian** [comment]
Yes, the exception was added in the very last version of the numpy.

**Comment by Tabrizian** [comment]
This is fixed on `master` branch. Feel free to reopen if you still see this issue.
